Restore the threaded graphics renderer.

pchote opened the pull request:
Closes #15211. ~~Depends on #15218.  Only the last two commits are new for this PR.~~

This PR restores the last commit from #15177, changed to target just the `IGraphicsContext` instead of the entire window.  The only real difference from what was merged last time is the second to commit that ensures thread safe access to the window size/scale properties.

RoosterDragon commented on OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:
```diff
@@ -0,0 +1,308 @@
+			Console.WriteLine("Using SDL 2 with OpenGL renderer");
+
+			// Lock the Window/Surface properties until initialization is complete
+			Monitor.Enter(syncObject);
```
Any reason to avoid the `lock` keyword here?

pchote committed:
Allow window size/scale properties to be accessed from other threads.

pchote commented:
Turns out that this final implementation doesn't have a problem with line length, so swapped the `Monitor.Enter`/`Exit` with a `lock`, fixed the `Sdl2HardwareCursor` disposal, and rebased.

RoosterDragon commented on OpenRA.Platforms.Default/Sdl2PlatformWindow.cs:
```diff
@@ -26,122 +27,159 @@ sealed class Sdl2PlatformWindow : ThreadAffine, IPlatformWindow
 		internal readonly IntPtr Window;
```
I think this technically needs to be covered by the `syncObject` too since it will be read from the render thread.

pchote replied:
Really? This is initialized before the renderer thread is created, and doesn't change.

RoosterDragon replied:
I'm not familiar enough with the threading model to *prove* reading it in this specific instance without any synchronisation is safe - even though I wouldn't be surprised if your reasoning is indeed accurate.

Unless I can prove something is safe - I'd prefer to just take a lock and remove any doubt. It doesn't look likely to be a problem for performance here.

pchote replied:
I agree, better safe than sorry.

RoosterDragon commented on OpenRA.Platforms.Default/ThreadedGraphicsContext.cs:
```diff
@@ -0,0 +1,702 @@
+			this.batchSize = batchSize;
+			renderThread = new Thread(RenderThread)
+			{
+				Name = "ThreadedGraphicsDevice RenderThread",
```
Should now be `Name = "ThreadedGraphicsContext RenderThread"`. (Could use `typeof` if you like - up to you).

RoosterDragon commented:
Confirmed that:

- We are still seeing the performance improvement from before.
- We can start and cleanly exit on the `ThreadedGraphicsContext` without a crash.
- We can start and cleanly exit on the plain `Sdl2GraphicsContext` without a crash.
- #15191 is still fixed.
- #15189 is still fixed (pressing ESC a ton during loading no longer seems to cause a crash).
- #15188 is still fixed (inserting a dummy error causes it to crash in the expected way - rather than an unexpected way)

The comment on #15193 suggests that issue is also resolved by this. :)

#15215 suggests the render thread got stuck on a `WaitOne` call waiting for some message that was passed using `Send` to complete. That could certainly be caused by general threading issues that this PR is resolving. But it may not. Might be worth keeping that one open for a bit when we merge this to see if it reoccurs or not.

LGTM - but since I wrote some of it I'll still suggest two other people weigh in on the PR.

RoosterDragon committed:
Run graphics rendering on a dedicated thread.

The main game thread can offload some of the CPU cost to the rendering thread, freeing up its time to run more logic and render ticks.

pchote commented:
Updated.

reaperrr commented:
> LGTM - but since I wrote some of it I'll still suggest two other people weigh in on the PR.

Tbh, out of the reviewers left I doubt anyone other than you and @chrisforbes would be able to spot any remaining conceptual issues.

Didn't encounter any problems while testing. I'll leave this unmerged for a bit in case @chrisforbes wants to take a look.

abcdefg30 commented:
> Didn't encounter any problems while testing.

Same. 👍 / ✅

GraionDilach commented on OpenRA.Platforms.Default/OpenRA.Platforms.Default.csproj:
```diff
@@ -2,7 +2,7 @@
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
```
Wouldn't this cause issues with macOS pushing out x86?

pchote commented:
Fixed.

GraionDilach commented:
I can't see anything bad in the code either (although I do feel I don't know half of the things I'm looking at now), and it works here as well, so... ✅ from me.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/6e1cde001aa9ef6439294b5544cefe0eebeb0a4d
